### PR TITLE
Feature/eslint

### PR DESCRIPTION
**Pull request by klaasman**

### CODE REVIEW

File: library/webpack-plugins/copy-unused-files-plugin.js [hunk 1]
```diff
@@ -54,14 +54,14 @@ function copyUnusedFilesPlugin() {
...
   return new Promise((resolve, reject) => {
     fs.copy(source, target, err => { err ? reject(err) : resolve() })
   })
 }
 
-function stat(filePath) {
+function stat (filePath) {
   return new Promise((resolve, reject) => {
-    fs.stat(filePath, (err, stats) => { err ? reject(er) : resolve(stats) })
+    fs.stat(filePath, (err, stats) => { err ? reject(err) : resolve(stats) })
```
Comment (klaasman) [inline library/webpack-plugins/copy-unused-files-plugin.js]: Linting caught at least one potentional bug :-)

**Comment by EECOLOR** [comment]
First of all: great work!

I'm afraid to say this: there is a few linting issues I don't agree with. One of them is the space after the function name, I really hate that convention. I like to copy functions before I call them and I hate to remove the space in that case. On top of that, it just does not make sense to me to have a difference between calling it and defining it.

There is a few other conventions I don't agree with, but I will comment on them inline.

**Comment by klaasman** [comment]
Don’t be afraid, I was expecting your opinion :-) we’ll discuss it tomorrow.

File: library/webpack-plugins/config-loader-plugin.js [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
       compiler.plugin('normal-module-factory', normalModuleFactory => {
 
         normalModuleFactory.plugin('after-resolve', (data, done) => {
-          const { loaders, rawRequest } = data
-          
-          if (data.rawRequest === '@kaliber/config')
-            loaders.push({ loader: require.resolve('../webpack-loaders/config-loader') })
+          const { loaders } = data
+
+          if (data.rawRequest === '@kaliber/config') { loaders.push({ loader: require.resolve('../webpack-loaders/config-loader') }) }
```
Comment (EECOLOR) [inline library/webpack-plugins/config-loader-plugin.js]: Two options:
1. Remove `data.` and put `rawRequest` back in the destructuring thing
2. Remove the destructuring and add `data.` to the loaders

File: library/webpack-plugins/copy-unused-files-plugin.js (hunk 1 above)
Comment (EECOLOR) [reply library/webpack-plugins/copy-unused-files-plugin.js]: :+1:

File: library/webpack-plugins/merge-css-plugin.js [hunk 3]
```diff
@@ -64,8 +64,8 @@ module.exports = function mergeCssPlugin() {
         // make sure the __webpack_css_chunk_hash__ is available in modules (code copied from ExtendedApiPlugin)
         compilation.dependencyFactories.set(ConstDependency, new NullFactory())
         compilation.dependencyTemplates.set(ConstDependency, new ConstDependency.Template())
-        compilation.mainTemplate.plugin('require-extensions', function(source, chunk, hash) {
-          const cssHash = (x => x && x.cssHash || 'no_css_files_in_chunk')(chunkCssAssets[chunk.name])
+        compilation.mainTemplate.plugin('require-extensions', function (source, chunk, hash) {
```
Comment (EECOLOR) [inline library/webpack-plugins/merge-css-plugin.js]: we could remove the function keyword here

File: library/webpack-plugins/merge-css-plugin.js [hunk 4]
```diff
@@ -84,20 +84,23 @@ module.exports = function mergeCssPlugin() {
         compilation.plugin('additional-chunk-assets', (chunks) => {
 
           const chunksByName = chunks.reduce(
-            (result, chunk) => (result[chunk.name] = chunk, result),
+            (result, chunk) => {
+              result[chunk.name] = chunk
+              return result
```
Comment (EECOLOR) [inline library/webpack-plugins/merge-css-plugin.js]: Hmm, I don't think this is an improvement.

File: library/webpack-plugins/react-universal-plugin.js [hunk 5]
```diff
@@ -147,11 +147,9 @@ function createWebCompiler(compiler, getEntries) {
     normalModuleFactory.plugin('after-resolve', (data, done) => {
       const { loaders, rawRequest, resourceResolveData: { query } } = data
 
-      if (query === '?universal-client')
-        loaders.push({ loader: require.resolve('../webpack-loaders/react-universal-client-loader') })
+      if (query === '?universal-client') { loaders.push({ loader: require.resolve('../webpack-loaders/react-universal-client-loader') }) }
```
Comment (EECOLOR) [inline library/webpack-plugins/react-universal-plugin.js]: Here I don't like the extra noise of the `{` and `}` and on top of that we should probably adjust our line length as well. I rather have that statement under the `if`, indented as it was.

File: library/webpack-plugins/websocket-communication-plugin.js [hunk 6]
```diff
@@ -45,7 +45,7 @@ function websocketCommunicationPlugin() {
       compiler.plugin('compilation', (compilation, { normalModuleFactory }) => {
         compilation.dependencyFactories.set(ConstDependency, new NullFactory())
         compilation.dependencyTemplates.set(ConstDependency, new ConstDependency.Template())
-        compilation.mainTemplate.plugin('require-extensions', function(source, chunk, hash) {
+        compilation.mainTemplate.plugin('require-extensions', function (source, chunk, hash) {
```
Comment (EECOLOR) [inline library/webpack-plugins/websocket-communication-plugin.js]: Isn't there a rule that disallows anonymous `function` definitions? If we use them they should be given a name anyway.

Comment (klaasman) [reply library/webpack-plugins/websocket-communication-plugin.js]: Perhaps the [`func-style`](https://eslint.org/docs/rules/func-style) in combination with the `"allowArrowFunctions": true`-setting.

File: example/src/index.html.js [hunk 7]
```diff
@@ -39,6 +39,11 @@ function getMessage() {
 
 export default main
 
+main.propTypes = {
+  location: object,
+  data: object
+}
```
[the remark this answers is not in the document]
Comment (klaasman) [reply example/src/index.html.js]: Perhaps `skipUndeclared` can help us

File: library/webpack-loaders/css-loader.js [hunk 8]
```diff
@@ -46,8 +46,8 @@ module.exports = function CssLoader(source, map) {
   const filename = relative(this.options.context, this.resourcePath)
   const options = {
     from: this.resourcePath,
-    to  : this.resourcePath,
-    map : { prev: map || false, inline: false, annotation: false }
+    to: this.resourcePath,
+    map: { prev: map || false, inline: false, annotation: false }
```
[the remark this answers is not in the document]
Comment (klaasman) [reply library/webpack-loaders/css-loader.js]: solution: disable the rule

**Comment by klaasman** [comment]
Ok, so I've updated the ruleset to be less opinionated, based on the ruleset of `eslint-config-react-app`.

The lint output for the `library`-directory:

```
/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/bin/serve.js
  56:76  warning  Unexpected use of comma operator  no-sequences
  77:28  warning  Unexpected mix of '&&' and '||'   no-mixed-operators
  77:59  warning  Unexpected mix of '&&' and '||'   no-mixed-operators

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/lib/build.js
  275:55  warning  Unexpected use of comma operator  no-sequences

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/lib/hot-css-replacement-client.js
   7:10  warning  'hotCssReplacementClient' is defined but never used  no-unused-vars
  17:5   warning  Expected a default case                              default-case

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/lib/hot-module-replacement-client.js
  11:5   warning  Expected a default case            default-case
  13:22  error    '__webpack_hash__' is not defined  no-undef

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/lib/node-utils.js
  11:11  warning  'result' is assigned a value but never used              no-unused-vars
  11:20  warning  eval can be harmful                                      no-eval
  32:16  warning  Expected to return a value at the end of arrow function  array-callback-return
  33:40  warning  Arrow function expected a return value                   array-callback-return

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/lib/stylesheet.js
   5:1   error  Unexpected '!' in 'raw-loader!./hot-css-replacement-client'. Do not use import syntax to configure webpack loaders  import/no-webpack-loader-syntax
  10:19  error  '__webpack_css_chunk_hash__' is not defined                                                                         no-undef
  11:36  error  '__webpack_websocket_port__' is not defined                                                                         no-undef
  11:64  error  '__webpack_css_chunk_hash__' is not defined                                                                         no-undef
  11:92  error  '__webpack_chunkname__' is not defined                                                                              no-undef

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/postcss-plugins/postcss-url-replace.js
  25:9  warning  Unexpected assignment within a 'while' statement  no-cond-assign

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-loaders/fragment-loader.js
   5:7   warning  'fileLoader' is assigned a value but never used  no-unused-vars
  13:39  warning  Unexpected mix of '&&' and '||'                  no-mixed-operators
  13:85  warning  Unexpected mix of '&&' and '||'                  no-mixed-operators

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-loaders/react-universal-client-loader.js
  8:34  warning  Unnecessary escape character: \.  no-useless-escape

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-loaders/react-universal-server-loader.js
  9:34  warning  Unnecessary escape character: \.  no-useless-escape

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-loaders/template-loader.js
  7:9  warning  'relative' is assigned a value but never used  no-unused-vars

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-plugins/config-loader-plugin.js
  10:28  warning  'rawRequest' is assigned a value but never used  no-unused-vars

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-plugins/copy-unused-files-plugin.js
  31:64  warning  Arrow function expected a return value  array-callback-return
  65:54  error    'er' is not defined                     no-undef

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-plugins/merge-css-plugin.js
   68:35  warning  Unexpected mix of '&&' and '||'      no-mixed-operators
   68:48  warning  Unexpected mix of '&&' and '||'      no-mixed-operators
   87:59  warning  Unexpected use of comma operator     no-sequences
   94:45  warning  Unnecessary escape character: \.     no-useless-escape
  100:29  warning  Expected '!==' and instead saw '!='  eqeqeq

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-plugins/template-plugin.js
   27:33  warning  Unnecessary escape character: \.     no-useless-escape
   58:35  warning  Expected '!==' and instead saw '!='  eqeqeq
   87:59  warning  Unexpected use of comma operator     no-sequences
  114:32  warning  Expected '!==' and instead saw '!='  eqeqeq

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/library/webpack-resolver-plugins/fragment-resolver-plugin.js
  8:47  warning  Unexpected mix of '&&' and '||'  no-mixed-operators
  8:74  warning  Unexpected mix of '&&' and '||'  no-mixed-operators

✖ 38 problems (7 errors, 31 warnings)
```

The lint output for the `example`-directory:
```
/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/example/src/index.html.js
  64:9  warning  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text
  65:9  warning  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text
  66:9  warning  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/example/src/partials/head.js
  1:8  warning  'styles' is defined but never used  no-unused-vars
  2:8  warning  'x' is defined but never used       no-unused-vars
  3:8  warning  'y' is defined but never used       no-unused-vars

/Users/klandsman/Sites/kaliber/kaliber/kaliber-build/example/src/test/index.html.js
  2:8  warning  'styles' is defined but never used  no-unused-vars

✖ 7 problems (0 errors, 7 warnings)

✨  Done in 0.95s.
```

**Comment by EECOLOR** [comment]
Below my opinion

Keep the rule (prevent the usage in code):
- Unexpected mix of '&&' and '||'   no-mixed-operators
- 'hotCssReplacementClient' is defined but never used  no-unused-vars
- Expected a default case                              default-case
-  '__webpack_hash__' is not defined  no-undef
- eval can be harmful                                      no-eval
- Unexpected '!' in 'raw-loader!./hot-css-replacement-client'. Do not use import syntax to configure webpack loaders  import/no-webpack-loader-syntax
- Unnecessary escape character: \.  no-useless-escape
- Expected '!==' and instead saw '!='  eqeqeq
- img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text

Disable the rule (allow the usage in code):
- Unexpected use of comma operator  no-sequences
- Expected to return a value at the end of arrow function  array-callback-return
-  Arrow function expected a return value                   array-callback-return
- Unexpected assignment within a 'while' statement  no-cond-assign

`array-callback-return` is one that I am willing to move to the 'keep' list. But it greatly depends on how often people actually forget a return. I use it sometimes (as you can see). I wonder if there is a rule that forces the use of `{` and `}` in `forEach` arrow functions, anyway, that's off topic.

**Comment by klaasman** [comment]
I'll only remove the `no-sequences` rule.
- `array-callback-return`: can catch potential bugs
- `no-cond-assign`: can catch potential bugs: it's not only applicable to your `while`-loop, but also to every `if`-statement

**Pull request merged**